### PR TITLE
Reduce the friction in getting these engines talking to each other.

### PR DESCRIPTION
**Pull request by eric1234**
Reduces the friction for getting these two engines working together. [The commit message](https://github.com/AlchemyCMS/alchemy-solidus/commit/325f637944484e02fb724d4373e15dcf8ac0f027) includes the full details. With these changes to the setup issues like #4 will be resolved.

### CODE REVIEW

**Comment by ridem** [comment]
This commit was the only way for me to get around issue #4 . Would be awesome to have it in `master` so I can remove my hacks

**Comment by tvdeyen** [comment]
@eric1234 @ridem this is lot of stuff to get my head around to. We need to assure that it works for all not only for specific cases. So, before we hastily merge this into master we need to review it thoroughly.

@eric1234 could you please split up 325f637944484e02fb724d4373e15dcf8ac0f027 into smaller commits. Maybe one for each headline? That would be far easier to review. Thanks.

**Comment by tvdeyen** [comment]
Side note: We discussed the `alias_action` thing in Solidus Slack lately. So maybe @senjai can add some thoughts?

**Comment by tvdeyen** [comment]
Some additional thoughts:

1. We definitely should remove the [auto-magically merging `::Ability` classes](https://github.com/AlchemyCMS/alchemy_cms/blob/master/lib/alchemy/controller_actions.rb#L71-L73) from Alchemy core.
2. I don't think we really need to merge the abilities at all. Where ever do we really have both engines involved in the same context (views, controllers, helpers?) We should consider to have to collections on each user `alchemy_roles` and `spree_roles` instead, so a user could have rights in Alchemy and in Spree. A user could even be a "Shop admin" but not an "CMS editor".

Any thoughts @mamhoff ?

**Comment by tvdeyen** [comment]
Cross referencing https://github.com/AlchemyCMS/alchemy_cms/issues/1022

File: README.md [hunk 1]
```diff
@@ -149,28 +145,68 @@ $ bundle exec rake alchemy:install
...
-...
-Spree::UserSessionsController.class_eval do
-  include Alchemy::ControllerActions
-end
+# config/routes.rb
+mount Spree::Core::Engine => '/store'
+mount Alchemy::Engine => '/pages'
+```
+
+### Put Soidus at the root level and Alchemy in it's own namespace:
```
Comment (mamhoff) [inline README.md]: `Solidus` typo

File: README.md [hunk 2]
```diff
@@ -149,28 +145,68 @@ $ bundle exec rake alchemy:install
...
+
+```ruby
+# config/routes.rb
+mount Alchemy::Engine => '/pages'
+mount Spree::Core::Engine => '/'
+```
+
+### Put Alchemy at the root level and Solidus in it's own namespace:
```
Comment (ridem) [inline README.md]: `it's` typo

File: README.md (hunk 1 above)
Comment (ridem) [inline README.md]: `it's` typo

**Comment by mamhoff** [comment]
We just removed the auto-magic merging of abilities. For all the other stuff, I need to think about this more properly. It seems like this commit is less explicit than I like it. I'd rather have the `Alchemy::ControllerHelpers` properly namespace their object references, for example, rather than creating a copy that lives in the same namespace. But: I might think differently about this tomorrow, so bear with me for a day or two :)

**Comment by eric1234** [comment]
I have split my giant multi-focused commit into hopefully smaller more focused commits which should make it easier to discuss. A few notes:

1. The existing system already had `alchemy_roles` and `spree_roles` methods and I didn't mess with that (other than making it easier and reload more reliably).
2. The core problem is that cancancan calls `current_ability` to get the `Ability` object which abides by those roles. The problem is that depending on our context we want either the Spree ability or the Alchemy ability. Without any integration this works out-of-the-box. But if we include the Alchemy controller actions into Spree's controller then Alchemy's `current_ability` ends up overriding Spree's. My fix masks Alchemy's `current_ability` (getting everything but that). A cleaner solution might be to refactor Alchemy so that there is a module of all actions except `current_ability` (that systems like Spree can use) and then another module which Alchemy uses that defines the `current_ability` as well as includes the same module everybody else is using. I.E. Alchemy provides a module that doesn't need to be cleaned to avoid the clash.

**Comment by eric1234** [comment]
NOTE: For the most part my new commits should be the same as my original big commit but I may have missed something. I'm going to point my app to this new commit stream tomorrow to do a sanity test to gain confidence. Until then the split commits are useful for discussion but they haven't seen action yet so you probably don't want to merge quite yet if you do like them.

**Comment by tvdeyen** [comment]
Thank you. This is very much appreciated. I like the idea to extract the `current_ability` into its own module like Spree/Solidus does with the 'Spree::ControllerActions::Auth` module. Would you mind to open an issue at the Alchemy repo?

**Comment by eric1234** [comment]
To hopefully make this faster I just created a PR rather than an issue. Once it's adjusted however need be and is committed I'll refactor 018edc3c0 to be simpler.

**Comment by tvdeyen** [comment]
@eric1234 https://github.com/AlchemyCMS/alchemy_cms/pull/1028 is merged

Do you want to update this PR?

**Comment by eric1234** [comment]
Sure thing. Now that `Alchemy::ControllerActions` no longer adds the conflicting method it greatly simplifies what we are doing here. These commits still reduce the work in getting alchemy and solidus working together (simple requires), fixes some reloading issues with the user extensions and adds some additional integration documentation information but not as much in this commit anymore. :)

File: README.md [hunk 3]
```diff
@@ -149,28 +145,68 @@ $ bundle exec rake alchemy:install
...
+initializer:
 
 ```ruby
-# config/initializers/solidus.rb
-...
-Spree::BaseController.class_eval do
-  include Alchemy::ControllerActions
-end
+# config/initializers/alchemy.rb
+require 'alchemy/solidus/alchemy_in_solidus'
```
Comment (tvdeyen) [inline README.md]: This is the same as above. Is this intentional?

**Comment by tvdeyen** [comment]
@eric1234 great stuff! Thanks. I hope I can free some time this weekend and test this thoroughly, but I'm very confident it will.

Great work and thanks for sharing with the rest of us!

File: README.md (hunk 3 above)
Comment (eric1234) [inline README.md]: Good catch. When splitting the commits I must have gotten things moved around. I have uploaded new commits that should be right and still is spit up some to allow better review.

**Pull request merged**